elasticsearch_dsl/_async/search.py

```diff
@@ -0,0 +1,146 @@
+#  Licensed to Elasticsearch B.V. under one or more contributor
+#  license agreements. See the NOTICE file distributed with
+#  this work for additional information regarding copyright
+#  ownership. Elasticsearch B.V. licenses this file to you under
+#  the Apache License, Version 2.0 (the "License"); you may
+#  not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+# 	http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing,
+#  software distributed under the License is distributed on an
+#  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+#  KIND, either express or implied.  See the License for the
+#  specific language governing permissions and limitations
+#  under the License.
+
+from elasticsearch.exceptions import ApiError
+from elasticsearch.helpers import async_scan
+
+from ..async_connections import get_connection
+from ..response import Response
+from ..search_base import MultiSearchBase, SearchBase
+from ..utils import AttrDict
+
+
+class AsyncSearch(SearchBase):
+    def __aiter__(self):
+        """
+        Iterate over the hits.
+        """
+
+        class ResultsIterator:
```

Comment:
@pquentin Also would love your opinion on this iterator work. Due to using unasync, we're forced to have identical code for the async and sync sides, so I had to implement this iterator.

In reality there is no need for the iterator to be async even in the async side, since we retrieve all the results in a single call, but because the `execute()` call is async I didn't see any other choice other than making the whole iteration async.

Thoughts on this? I would have preferred to use a sync iterator for both async and sync (which I believe is what the main client does, right?), but I don't think we have a way to do that if we want the `AsyncSearch` object to be iterable.

Reply:
> In reality there is no need for the iterator to be async even in the async side, since we retrieve all the results in a single call, but because the `execute()` call is async I didn't see any other choice other than making the whole iteration async.

I can see why we would want a sync iterator and how making it async makes it look like we're maybe making multiple requests, yes. However, I think it's the current design of the library that forces our hand, not unasync. What we *could* do is encourage using `Response.__iter__` instead of `Search.__iter__` by documenting `for hit in Search(...).execute()` and `for hit in await AsyncSearch(...).execute()` instead of the current `for hit in Search(...)`. But on the other hand, `async for hit in AsyncSearch(...)` is not *that* bad?

Also, I think that allowing `Search.__iter__` but not `AsyncSearch.__iter__` would just be confusing. So in summary, I think `ResultsIterator` is fine.

> I would have preferred to use a sync iterator for both async and sync (which I believe is what the main client does, right?)

The main client does not expose iterators, except for helpers, which are async and sync.

Reply:
I think it looks nicer to iterate on the `AsyncSearch` and `Search` objects, without the explicit `execute()` call. So okay, I guess this is fine then.

Reply:
Yes, and using `async for` clarifies that I/O is in fact happening.